### PR TITLE
Add support for arbitrary writable volume mounts during `pack build`

### LINKED ISSUE #764: Update --volume to support arbitrary target locations
### Description

To enable further local experimentation and development for `pack` users of buildpacks, builders or lifecycles we would like to loosen the constraints on where volumes may be mounted with the possibility for those mounts to be writable.

Additionally, there has been feedback that the current implementation and it's automatic appending of `/platform` to the target location is surprising.

### Proposed solution

This issue is a request to implement proposed solution as detail in https://github.com/buildpacks/rfcs/pull/85.

Summary of changes:

1. Allow `--volume` to support any target location
2. Support a third colon mode parameter defaulting to `ro`. (eg. `--volume=/tmp/dir:/some-dir:rw`)
3. `WARN` when a spec'd directory (aside from `/platform`) is targeted.

### Additional context
- Blocked by https://github.com/buildpacks/rfcs/pull/85
- This feature should be documented as part of standard pack CLI commands documentation.

### PR DESCRIPTION
**Pull request by jromero**
## Summary
This PR adds support to volume mount at any location during `pack build`. It additionally adds support for writable mounts (`rw`).

##### `pack build --help`

```
$ ./out/pack build --help
Generate app image from source code

Usage:
  pack build <image-name> [flags]

Flags:
  -B, --builder string              Builder image (default "gcr.io/paketo-buildpacks/builder:base")
  -b, --buildpack strings           Buildpack reference in the form of '<buildpack>@<version>',
                                      path to a buildpack directory (not supported on Windows),
                                      path/URL to a buildpack .tar or .tgz file, or
                                      the name of a packaged buildpack image
                                    Repeat for each buildpack in order,
                                      or supply once by comma-separated list
  -R, --buildpack-registry string   Buildpack Registry URL
      --clear-cache                 Clear image's associated cache before building
  -D, --default-process string      Set the default process type
  -d, --descriptor string           Path to the project descriptor file
  -e, --env stringArray             Build-time environment variable, in the form 'VAR=VALUE' or 'VAR'.
                                    When using latter value-less form, value will be taken from current
                                      environment at the time this command is executed.
                                    This flag may be specified multiple times and will override
                                      individual values defined by --env-file.
      --env-file stringArray        Build-time environment variables file
                                    One variable per line, of the form 'VAR=VALUE' or 'VAR'
                                    When using latter value-less form, value will be taken from current
                                      environment at the time this command is executed
  -h, --help                        Help for 'build'
      --network string              Connect detect and build containers to network
      --no-pull                     Skip pulling builder and run images before use
  -p, --path string                 Path to app dir or zip-formatted file (defaults to current working directory)
      --publish                     Publish to registry
      --run-image string            Run image (defaults to default stack's run image)
      --trust-builder               Trust the provided builder
                                    All lifecycle phases will be run in a single container (if supported by the lifecycle).
      --volume stringArray          Mount host volume into the build container, in the form '<host path>:<target path>[:<mode>]'.
                                    Repeat for each volume in order,
                                      or supply once by comma-separated list

Global Flags:
      --no-color     Disable color output
  -q, --quiet        Show less output
      --timestamps   Enable timestamps in output
  -v, --verbose      Show more output
```

##### mounting a custom lifecycle

```
$ pack build my-bash-app -B cnbs/sample-builder:bionic --volume=/Users/javier.romero/dev/buildpacks/lifecycle/out/linux/lifecycle/:/cnb/lifecycle/ -p ~/dev/buildpacks/samples/apps/bash-script/
Warning: Using untrusted builder with volume mounts. If there is sensitive data in the volumes, this may present a security vulnerability.
bionic: Pulling from cnbs/sample-builder
Digest: sha256:3665853288be6c7df936f53dd8fccea10d22785c3015aec4dd49ab44f9c3a7bb
Status: Image is up to date for cnbs/sample-builder:bionic
bionic: Pulling from cnbs/sample-stack-run
Digest: sha256:4c51a26b0546508ccd6f2d14753d041b595b3f9bceef77f4bb969e06c66474a4
Status: Image is up to date for cnbs/sample-stack-run:bionic
Warning: Mounting to a sensitive directory /cnb/lifecycle/
0.8.0: Pulling from buildpacksio/lifecycle
Digest: sha256:48dfb79e342fdeb68a1bf310b33b349269b2919cb5029e0b7184b84e82fc0bb3
Status: Image is up to date for buildpacksio/lifecycle:0.8.0
===> DETECTING
[detector] samples/bash-script 0.0.1
===> ANALYZING
[analyzer] Previous image with name "index.docker.io/library/my-bash-app:latest" not found
===> RESTORING
===> BUILDING
[builder] ---> Hello World buildpack
[builder] ---> Hello Bash Script buildpack
[builder] 
[builder] Here are the contents of the current working directory:
[builder] .:
[builder] total 16
[builder] drwxr-xr-x 2 cnb  cnb  4096 Aug  5 18:50 .
[builder] drwxr-xr-x 1 root root 4096 Aug  5 18:50 ..
[builder] -rwxr-xr-x 1 cnb  cnb   738 Apr  6 23:13 app.sh
[builder] -rw-r--r-- 1 cnb  cnb   202 Apr  6 23:13 project.toml
===> EXPORTING
[exporter] Adding layer 'launcher'
[exporter] Adding 1/1 app layer(s)
[exporter] Adding layer 'config'
[exporter] *** Images (d1b9f852dd37):
[exporter]       index.docker.io/library/my-bash-app:latest
Successfully built image my-bash-app
```

## Documentation

- Should this change be documented?
    - [x] Yes, see https://github.com/buildpacks/docs/issues/181
    - [ ] No

## Related

Resolves #764

### CODE REVIEW

File: acceptance/acceptance_test.go [hunk 1]
```diff
@@ -884,55 +884,90 @@ func testAcceptance(
...
+							readWriteBuildpackTgz,
+							tmpVolumeSrc string
+						)
 
 						it.Before(func() {
 							h.SkipUnless(t,
-								pack.SupportsFeature(invoke.CustomVolumeMounts),
-								"pack 0.11.0 shipped with a volume mounting bug",
+								pack.SupportsFeature(invoke.ReadWriteVolumeMounts),
+								"pack version does not support volume mounts",
```
Comment (jromero) [inline acceptance/acceptance_test.go:896]:
```suggestion
								"pack version does not support read/write volume mounts",
```

File: acceptance/invoke/pack.go [hunk 2]
```diff
@@ -250,25 +246,18 @@ func (i *PackInvoker) semanticVersion() *semver.Version {
...
-}
-
-func (i *PackInvoker) laterThan0_10_0() bool {
-	ver := i.semanticVersion()
-	return ver.GreaterThan(semver.MustParse("0.10.0")) || ver.Equal(semver.MustParse("0.0.0"))
-}
-
-func (i *PackInvoker) not0_11_0() bool {
+// laterThan returns true if pack version is older than the provided version
+func (i *PackInvoker) laterThan(version string) bool {
```
Comment (dfreilich) [inline acceptance/invoke/pack.go:250]: 👍

**Comment by jromero** [comment]
> Looks great. Should we have a warning that this may not work on other OSs?

The only one I know of in which this _may_ not work is Windows (`WCOW`) but honestly haven't thoroughly tested it in `LCOW`.

**Comment by ameyer-pivotal** [comment]
> > Looks great. Should we have a warning that this may not work on other OSs?
> 
> The only one I know of in which this _may_ not work is Windows (`WCOW`) but honestly haven't thoroughly tested it in `LCOW`.

For WCOW, the solution would be fairly simple. Just use the `mounts.NewParser(mounts.OSWindows)` parser and users would just need to provide the target location as `c:/path/in/container`.

**Comment by jromero** [comment]
> > > Looks great. Should we have a warning that this may not work on other OSs?
> > 
> > 
> > The only one I know of in which this _may_ not work is Windows (`WCOW`) but honestly haven't thoroughly tested it in `LCOW`.
> 
> @jromero for WCOW, the solution would be fairly simple. Just use the `mounts.NewParser(mounts.OSWindows)` parser and users would just need to provide the target location as `c:/path/in/container`.

@ameyer-pivotal Thank you. If I'm not mistaken this will be taken care of as part of https://github.com/buildpacks/pack/pull/739

**Pull request merged**